sklearn/preprocessing/data.py

```diff
@@ -1904,44 +1904,75 @@ def transform(self, X):
                                    self.categorical_features, copy=True)
 
 
+def _get_count_dict_3():
+    """Gets the outer count dictionary."""
+
+    # get_count_dict() is an optimization on key-value access
+    # where we have multi-layer dicts for accessing dict[X][y] instead of
+    # using dict[(X, y)], because using dict[(X, y)] is inefficient
+    # Using external helper methods is also required to avoid a picking
+    # error involving lambda expressions in defaultdict
+    return defaultdict(_get_count_dict_2)
+
+
+def _get_count_dict_2():
+    """Gets the inner count dictionary."""
+    return defaultdict(_get_count_dict_1)
+
+
+def _get_count_dict_1():
+    """Gets the inner count dictionary."""
+    return defaultdict(_get_count_dict_0)
+
+
+def _get_count_dict_0():
+    """Gets the innermost count dictionary."""
+    return defaultdict(int)
```

Comment:
Yes. I agree, we should experiment with data structures to find the right one. 
Currently I am leaning towards just making my own "default" dict that checks `if key in my_dict:` every time. 
I did a benchmark on this:

```
import time
from collections import defaultdict
from collections import Counter

loop_count = 10000000

t = time.time()
a = {}
for i in xrange(loop_count):
	if i in a:
		a[i] += i
	else:
		a[i] = i

for i in xrange(loop_count / 2):
	if i in a:
		a[i] += i
	else:
		a[i] = i

for i in xrange(loop_count * 2):
	if i in a:
		x = a[i] + 5
	else:
		x = 5

print "time take for {}:", (time.time() - t)


t = time.time()
b = defaultdict(int)
for i in xrange(loop_count):
	b[i] += i

for i in xrange(loop_count / 2):
	b[i] += i

for i in xrange(loop_count * 2):
	x = b[i] + 5

print "time take for defaultdict:", (time.time() - t)

t = time.time()
c = Counter(xrange(loop_count)) + Counter(xrange(loop_count / 2))

for i in xrange(loop_count * 2):
	x = c[i] + 5

print "time take for counter:", (time.time() - t)


t = time.time()
c2 = Counter()
for i in xrange(loop_count):
	c2[i] += i

for i in xrange(loop_count / 2):
	c2[i] += i

for i in xrange(loop_count * 2):
	x = c2[i] + 5

print "time take for counter version 2:", (time.time() - t)
```

And the results were kind of surprising:
```
time take for {}: 7.626983881
time take for defaultdict: 9.94141697884
time take for counter: 58.303508997
time take for counter version 2: 15.7543189526
```